File: python/devap/quality_gate.py
"""
Quality Gate — 品質閘門

依序執行驗證步驟：verify → lint → type_check → static_analysis → completion_criteria。
收集驗證證據（借鑑 Superpowers Iron Law）。

靜態合規檢查（非阻塞，僅 warning）：
- AGENTS.md 同步檢查
- 前端設計合規性檢查（XSPEC-026 AC-3.x）
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Awaitable, Callable, Optional

from devap.models.types import QualityConfig, Task, VerificationEvidence


@dataclass
class ShellResult:
    """Shell 執行結果"""

    exit_code: int
    stdout: str
    stderr: str


# Shell 執行器類型
ShellExecutor = Callable[[str, str], Awaitable[ShellResult]]


@dataclass
class QualityGateStep:
    """Quality Gate 單一步驟結果"""

    name: str
    command: str
    passed: bool
    output: str


@dataclass
class QualityGateResult:
    """Quality Gate 檢查結果"""

    passed: bool
    steps: list[QualityGateStep] = field(default_factory=list)
    feedback: Optional[str] = None
    evidence: list[VerificationEvidence] = field(default_factory=list)


@dataclass
class FrontendDesignCheckResult:
    """前端設計合規性檢查結果"""

    step: QualityGateStep
    missing_sections: Optional[list[str]] = None
    missing_color_tokens: Optional[list[str]] = None
    anti_pattern_count: Optional[int] = None


async def run_quality_gate(
    task: Task,
    quality_config: QualityConfig,
    cwd: str,
    shell_executor: ShellExecutor,
    on_progress: Optional[Callable[[str], None]] = None,
) -> QualityGateResult:
    """
    執行品質閘門

    Args:
        task: 要驗證的任務
        quality_config: 品質設定
        cwd: 工作目錄
        shell_executor: Shell 執行器
        on_progress: 進度回呼

    Returns:
        品質閘門檢查結果
    """
    steps: list[QualityGateStep] = []
    evidence: list[VerificationEvidence] = []

    has_test_levels = bool(task.test_levels and len(task.test_levels) > 0)

    async def run_step(name: str, command: str) -> QualityGateStep:
        step = await _execute_step(name, command, cwd, shell_executor)
        steps.append(step)
        evidence.append(VerificationEvidence(
            command=command,
            exit_code=0 if step.passed else 1,
            output=step.output[:2000],
            timestamp=datetime.now(timezone.utc).isoformat(),
        ))
        return step

    # 多層級測試模式
    if has_test_levels:
        for level in task.test_levels:  # type: ignore[union-attr]
            on_progress and on_progress(
                f"[{task.id}] Quality Gate: {level.name} → {level.command}"
            )
            step = await run_step(level.name, level.command)
            if not step.passed:
                return _build_fail_result(steps, step, evidence)

    elif quality_config.verify and task.verify_command:
        on_progress and on_progress(
            f"[{task.id}] Quality Gate: verify → {task.verify_command}"
        )
        step = await run_step("verify", task.verify_command)
        if not step.passed:
            return _build_fail_result(steps, step, evidence)

    # lint
    if quality_config.lint_command:
        on_progress and on_progress(
            f"[{task.id}] Quality Gate: lint → {quality_config.lint_command}"
        )
        step = await run_step("lint", quality_config.lint_command)
        if not step.passed:
            return _build_fail_result(steps, step, evidence)

    # type_check
    if quality_config.type_check_command:
        on_progress and on_progress(
            f"[{task.id}] Quality Gate: type_check → {quality_config.type_check_command}"
        )
        step = await run_step("type_check", quality_config.type_check_command)
        if not step.passed:
            return _build_fail_result(steps, step, evidence)

    # static_analysis
    if quality_config.static_analysis_command:
        on_progress and on_progress(
            f"[{task.id}] Quality Gate: static_analysis → {quality_config.static_analysis_command}"
        )
        step = await run_step("static_analysis", quality_config.static_analysis_command)
        if not step.passed:
            return _build_fail_result(steps, step, evidence)

    # completion_criteria
    if quality_config.completion_criteria:
        for check in quality_config.completion_criteria:
            if not check.command:
                continue
            on_progress and on_progress(
                f"[{task.id}] Quality Gate: completion_check → {check.command}"
            )
            step = await run_step("completion_check", check.command)
            if not step.passed and check.required:
                return _build_fail_result(steps, step, evidence)

    # 前端設計合規性檢查（非阻塞，僅 warning）
    # 純後端 API 或 CLI 專案若無 DESIGN.md，不視為錯誤
    fd_result = check_frontend_design_compliance(cwd)
    if fd_result is not None:
        steps.append(fd_result.step)
        evidence.append(VerificationEvidence(
            command="frontend_design_check",
            exit_code=0 if fd_result.step.passed else 1,
            output=fd_result.step.output[:2000],
            timestamp=datetime.now(timezone.utc).isoformat(),
        ))
        if not fd_result.step.passed:
            on_progress and on_progress(
                f"[{task.id}] Quality Gate: DESIGN.md compliance issues detected (warning only)"
            )

    return QualityGateResult(passed=True, steps=steps, evidence=evidence)


async def _execute_step(
    name: str,
    command: str,
    cwd: str,
    shell_executor: ShellExecutor,
) -> QualityGateStep:
    """執行單一品質檢查步驟"""
    try:
        result = await shell_executor(command, cwd)
        passed = result.exit_code == 0
        output = result.stdout if passed else result.stderr or result.stdout
        return QualityGateStep(
            name=name,
            command=command,
            passed=passed,
            output=output,
        )
    except Exception as e:
        return QualityGateStep(
            name=name,
            command=command,
            passed=False,
            output=str(e),
        )


def _build_fail_result(
    steps: list[QualityGateStep],
    failed_step: QualityGateStep,
    evidence: list[VerificationEvidence],
) -> QualityGateResult:
    """構建失敗結果"""
    return QualityGateResult(
        passed=False,
        steps=steps,
        feedback=(
            f"Quality Gate 失敗於 {failed_step.name} 步驟\n"
            f"指令: {failed_step.command}\n"
            f"輸出: {failed_step.output[:500]}"
        ),
        evidence=evidence,
    )


# ─────────────────────────────────────────────────────────────────────────────
# 前端設計合規性驗證
# 來源：UDS frontend-design-standards.ai.yaml（XSPEC-026 Phase 1）
# ─────────────────────────────────────────────────────────────────────────────

# DESIGN.md 必填段落（snake_case 和 kebab-case 兩種格式）
_REQUIRED_DESIGN_SECTIONS: list[list[str]] = [
    ["visual_theme", "visual-theme"],
    ["color_palette", "color-palette"],
    ["typography"],
    ["component_styling", "component-styling"],
    ["layout_spacing", "layout-spacing"],
    ["design_guidelines", "design-guidelines"],
]

# 語義色彩必要 token（snake_case 和 kebab-case 兩種格式）
_REQUIRED_COLOR_TOKENS: list[list[str]] = [
    ["background"],
    ["surface"],
    ["primary_text", "primary-text"],
    ["muted_text", "muted-text"],
    ["accent"],
]

# anti_patterns 建議最低條目數
_MIN_ANTI_PATTERN_COUNT = 5


def check_frontend_design_compliance(cwd: str) -> Optional[FrontendDesignCheckResult]:
    """
    驗證前端設計合規性（XSPEC-026 AC-3.x）。

    非阻塞：DESIGN.md 不存在時回傳 None（純後端專案不應被 block）。
    DESIGN.md 存在時驗證：
    1. 6 個必填段落是否完整（缺失 → error，step.passed=False）
    2. 語義色彩 5 個必要 token 是否存在（缺失 → warn）
    3. anti_patterns 條目數是否 ≥ 5（不足 → warn）

    Args:
        cwd: 專案根目錄路徑

    Returns:
        FrontendDesignCheckResult 或 None（若無 DESIGN.md）
    """
    design_md_path = Path(cwd) / "DESIGN.md"

    if not design_md_path.exists():
        # DESIGN.md 不存在 → 跳過（非前端專案不應被 block）
        return None

    content = design_md_path.read_text(encoding="utf-8")

    issues: list[str] = []
    missing_sections: list[str] = []
    missing_color_tokens: list[str] = []
    anti_pattern_count: Optional[int] = None

    # 1. 必填段落完整性檢查
    for aliases in _REQUIRED_DESIGN_SECTIONS:
        found = any(
            re.search(
                rf"(?:^#{'{1,6}'}\s+{re.escape(alias)}\b|^{re.escape(alias)}\s*:)",
                content,
                re.IGNORECASE | re.MULTILINE,
            )
            for alias in aliases
        )
        if not found:
            missing_sections.append(aliases[0])

    if missing_sections:
        issues.append(
            f"缺少必填段落（{len(missing_sections)} 個）：{', '.join(missing_sections)}\n"
            f"  → 請在 DESIGN.md 中補充上述段落（支援 ## section_name 或 section_name: 格式）"
        )

    # 2. 語義色彩 token 檢查（warn）
    for aliases in _REQUIRED_COLOR_TOKENS:
        found = any(
            re.search(rf"\b{re.escape(alias)}\s*[=:\-]", content, re.IGNORECASE)
            for alias in aliases
        )
        if not found:
            missing_color_tokens.append(aliases[0])

    if missing_color_tokens:
        issues.append(
            f"語義色彩 token 不完整，缺少：{', '.join(missing_color_tokens)}\n"
            f"  → 建議在 color_palette 段落中補充上述 token"
        )

    # 3. anti_patterns 條目數檢查（warn）
    ap_match = re.search(
        r"(?:anti[_\-]patterns?)\s*[:\n](.*?)(?=\n#{1,6}\s|\n\w[\w\-]*\s*:|$)",
        content,
        re.IGNORECASE | re.DOTALL,
    )
    if ap_match:
        list_items = re.findall(r"^\s*[-*+]\s+\S", ap_match.group(1), re.MULTILINE)
        anti_pattern_count = len(list_items)
        if anti_pattern_count < _MIN_ANTI_PATTERN_COUNT:
            issues.append(
                f"design_guidelines.anti_patterns 條目不足"
                f"（目前 {anti_pattern_count} 條，建議 ≥ {_MIN_ANTI_PATTERN_COUNT} 條）\n"
                f"  → 請補充常見的前端設計反模式"
            )

    passed = len(missing_sections) == 0
    output_lines: list[str] = []

    if passed and not issues:
        output_lines.append("DESIGN.md 前端設計合規性驗證通過。")
    else:
        output_lines.append("DESIGN.md 前端設計合規性問題：")
        for i, issue in enumerate(issues, 1):
            output_lines.append(f"\n[{i}] {issue}")
        if passed:
            output_lines.append("\n（必填段落完整，上述為 warning 項目）")
        else:
            output_lines.append("\n請修正上述 error 項目（必填段落缺失），warning 項目建議補充。")

    return FrontendDesignCheckResult(
        step=QualityGateStep(
            name="frontend_design_check",
            command="frontend_design_check",
            passed=passed,
            output="".join(output_lines),
        ),
        missing_sections=missing_sections or None,
        missing_color_tokens=missing_color_tokens or None,
        anti_pattern_count=anti_pattern_count,
    )

File: packages/core/src/quality-gate.ts
/**
 * Quality Gate — 品質門檻執行器
 *
 * 支援兩種模式：
 * 1. 傳統模式：依序執行 verify_command → lint_command → type_check_command
 * 2. 多層級模式：依序執行 test_levels 定義的多個測試層級
 *
 * test_levels 優先於 verify_command。全部通過才視為品質門檻通過。
 * 透過回呼函式執行 shell 指令，方便測試 mock。
 */

import { readFile, readdir } from "node:fs/promises";
import { join } from "node:path";
import { createHash } from "node:crypto";
import type { QualityConfig, Task, CompletionCheck, VerificationEvidence } from "./types.js";

/**
 * Quality Gate 檢查結果
 */
export interface QualityGateResult {
  /** 是否全部通過 */
  passed: boolean;
  /** 各步驟結果 */
  steps: QualityGateStep[];
  /** 失敗的回饋訊息（用於 fix loop 注入 agent prompt） */
  feedback?: string;
  /** 驗證證據（借鑑 Superpowers Iron Law：Evidence before claims） */
  evidence: VerificationEvidence[];
  /** 規格品質評分（由 UDS checklist scoring 提供） */
  score?: number;
  /** 規格品質滿分（Standard mode = 10, Boost mode = 25） */
  max_score?: number;
}

/**
 * Quality Gate 單步結果
 */
export interface QualityGateStep {
  /** 步驟名稱 */
  name: "verify" | "lint" | "type_check" | "static_analysis" | "completion_check"
       | "unit" | "integration" | "system" | "e2e" | "agents_md_check"
       | "frontend_design_check";
  /** 執行的指令 */
  command: string;
  /** 是否通過 */
  passed: boolean;
  /** 指令輸出（stdout + stderr） */
  output: string;
}

/**
 * 前端設計合規性檢查結果
 *
 * 包含缺失的必填段落與警告訊息。
 */
export interface FrontendDesignCheckResult {
  /** 檢查步驟結果 */
  step: QualityGateStep;
  /** 缺失的必填段落列表 */
  missingSections?: string[];
  /** 缺失的語義色彩 token 列表 */
  missingColorTokens?: string[];
  /** 反模式條目不足（實際數量） */
  antiPatternCount?: number;
}

/**
 * Shell 指令執行器回呼
 *
 * @param command - 要執行的 shell 指令
 * @param cwd - 工作目錄
 * @returns { exitCode, stdout, stderr }
 */
export type ShellExecutor = (
  command: string,
  cwd: string,
) => Promise<{ exitCode: number; stdout: string; stderr: string }>;

/**
 * Hook telemetry 記錄
 *
 * 由 adapter 在 agent 執行期間收集，
 * 記錄 PostToolUse hooks 已成功執行的品質檢查步驟。
 */
export interface HookTelemetry {
  /** lint hook 最後一次執行結果（true = 全部通過） */
  lint_passed?: boolean;
  /** type_check hook 最後一次執行結果（true = 全部通過） */
  type_check_passed?: boolean;
}

/**
 * Quality Gate 選項
 */
export interface QualityGateOptions {
  /** 工作目錄 */
  cwd: string;
  /** Shell 指令執行器 */
  shellExecutor: ShellExecutor;
  /** 進度回呼 */
  onProgress?: (message: string) => void;
  /** Hook telemetry（若有，跳過已通過的步驟） */
  hookTelemetry?: HookTelemetry;
}

/**
 * 執行品質門檻檢查
 *
 * 若 task 有 test_levels 且非空，走多層級模式（取代 verify_command）：
 *   依序執行每個 test level，第一個失敗即停止。
 *   之後再執行 lint_command、type_check_command。
 *
 * 否則走傳統模式：
 *   1. verify_command（若 task 有設定且 quality.verify 為 true）
 *   2. lint_command（若 quality config 有設定）
 *   3. type_check_command（若 quality config 有設定）
 *
 * 任一步驟失敗即停止，回傳失敗結果。
 *
 * @param task - 已執行完成的任務
 * @param qualityConfig - 品質設定
 * @param options - 執行選項
 * @returns Quality Gate 結果
 */
export async function runQualityGate(
  task: Task,
  qualityConfig: QualityConfig,
  options: QualityGateOptions,
): Promise<QualityGateResult> {
  const steps: QualityGateStep[] = [];
  const evidence: VerificationEvidence[] = [];

  // 判斷走多層級或傳統模式
  const hasTestLevels = task.test_levels && task.test_levels.length > 0;

  /** 執行步驟並收集證據的輔助函式 */
  const runStep = async (
    name: QualityGateStep["name"],
    command: string,
  ): Promise<QualityGateStep> => {
    const step = await executeStep(name, command, options);
    steps.push(step);
    // 收集驗證證據（Iron Law: Evidence before claims）
    evidence.push({
      command,
      exit_code: step.passed ? 0 : 1,
      output: step.output.slice(0, 2000),
      timestamp: new Date().toISOString(),
    });
    return step;
  };

  if (hasTestLevels) {
    // 多層級測試模式：依序執行每個 test level
    for (const level of task.test_levels!) {
      options.onProgress?.(`[${task.id}] Quality Gate: ${level.name} → ${level.command}`);
      const step = await runStep(level.name, level.command);
      if (!step.passed) {
        return buildFailResult(steps, step, evidence, task);
      }
    }
  } else if (qualityConfig.verify && task.verify_command) {
    // 傳統模式：verify_command
    options.onProgress?.(`[${task.id}] Quality Gate: verify_command → ${task.verify_command}`);
    const step = await runStep("verify", task.verify_command);
    if (!step.passed) {
      return buildFailResult(steps, step, evidence, task);
    }
  }

  // lint_command（兩種模式都執行，telemetry pass 時跳過）
  if (qualityConfig.lint_command) {
    if (options.hookTelemetry?.lint_passed === true) {
      options.onProgress?.(`[${task.id}] Quality Gate: lint → skipped (hook telemetry pass)`);
      steps.push({
        name: "lint",
        command: qualityConfig.lint_command,
        passed: true,
        output: "Skipped: hook telemetry indicates pass",
      });
      evidence.push({
        command: qualityConfig.lint_command,
        exit_code: 0,
        output: "Skipped: hook telemetry indicates pass",
        timestamp: new Date().toISOString(),
      });
    } else {
      options.onProgress?.(`[${task.id}] Quality Gate: lint → ${qualityConfig.lint_command}`);
      const step = await runStep("lint", qualityConfig.lint_command);
      if (!step.passed) {
        return buildFailResult(steps, step, evidence, task);
      }
    }
  }

  // type_check_command（兩種模式都執行，telemetry pass 時跳過）
  if (qualityConfig.type_check_command) {
    if (options.hookTelemetry?.type_check_passed === true) {
      options.onProgress?.(`[${task.id}] Quality Gate: type_check → skipped (hook telemetry pass)`);
      steps.push({
        name: "type_check",
        command: qualityConfig.type_check_command,
        passed: true,
        output: "Skipped: hook telemetry indicates pass",
      });
      evidence.push({
        command: qualityConfig.type_check_command,
        exit_code: 0,
        output: "Skipped: hook telemetry indicates pass",
        timestamp: new Date().toISOString(),
      });
    } else {
      options.onProgress?.(`[${task.id}] Quality Gate: type_check → ${qualityConfig.type_check_command}`);
      const step = await runStep("type_check", qualityConfig.type_check_command);
      if (!step.passed) {
        return buildFailResult(steps, step, evidence, task);
      }
    }
  }

  // static_analysis_command（兩種模式都執行）
  if (qualityConfig.static_analysis_command) {
    options.onProgress?.(`[${task.id}] Quality Gate: static_analysis → ${qualityConfig.static_analysis_command}`);
    const step = await runStep("static_analysis", qualityConfig.static_analysis_command);
    if (!step.passed) {
      return buildFailResult(steps, step, evidence, task);
    }
  }

  // completion_criteria（逐項執行有 command 的完成準則）
  if (qualityConfig.completion_criteria && qualityConfig.completion_criteria.length > 0) {
    for (const criterion of qualityConfig.completion_criteria) {
      if (!criterion.command) continue; // 無 command → 由 Judge 審查，跳過
      options.onProgress?.(`[${task.id}] Quality Gate: completion_check(${criterion.name}) → ${criterion.command}`);
      const step = await runStep("completion_check", criterion.command);
      if (!step.passed && criterion.required) {
        return buildFailResult(steps, step, evidence, task);
      }
    }
  }

  // AGENTS.md 合規檢查（非阻塞，僅 warning）
  const agentsMdResult = await checkAgentsMdSync(options.cwd);
  if (agentsMdResult) {
    steps.push(agentsMdResult.step);
    evidence.push({
      command: "agents_md_check",
      exit_code: agentsMdResult.step.passed ? 0 : 1,
      output: agentsMdResult.step.output.slice(0, 2000),
      timestamp: new Date().toISOString(),
    });
    if (!agentsMdResult.step.passed) {
      options.onProgress?.(`[${task.id}] Quality Gate: AGENTS.md drift detected (warning only)`);
    }
  }

  // 前端設計合規性檢查（非阻塞，僅 warning）
  // 純後端 API 或 CLI 專案若無 DESIGN.md，不視為錯誤
  const frontendDesignResult = await checkFrontendDesignCompliance(options.cwd);
  if (frontendDesignResult) {
    steps.push(frontendDesignResult.step);
    evidence.push({
      command: "frontend_design_check",
      exit_code: frontendDesignResult.step.passed ? 0 : 1,
      output: frontendDesignResult.step.output.slice(0, 2000),
      timestamp: new Date().toISOString(),
    });
    if (!frontendDesignResult.step.passed) {
      options.onProgress?.(`[${task.id}] Quality Gate: DESIGN.md compliance issues detected (warning only)`);
    }
  }

  return {
    passed: true,
    steps,
    evidence,
    ...(task.spec_score != null && {
      score: task.spec_score,
      max_score: task.spec_max_score ?? (task.spec_score <= 10 ? 10 : 25),
    }),
  };
}

/**
 * 執行單一品質門檻步驟
 */
async function executeStep(
  name: QualityGateStep["name"],
  command: string,
  options: QualityGateOptions,
): Promise<QualityGateStep> {
  try {
    const { exitCode, stdout, stderr } = await options.shellExecutor(command, options.cwd);
    const output = [stdout, stderr].filter(Boolean).join("\n");
    return {
      name,
      command,
      passed: exitCode === 0,
      output: output.slice(0, 5000),
    };
  } catch (error) {
    return {
      name,
      command,
      passed: false,
      output: `執行錯誤：${error instanceof Error ? error.message : String(error)}`,
    };
  }
}

/**
 * 構建失敗結果，含 feedback 供 fix loop 使用
 */
/**
 * 檢查 AGENTS.md 與 .standards/ 的同步狀態
 *
 * 非阻塞：drift 只產生 warning，不阻擋 QualityGate。
 * 檢查項目：
 * 1. AGENTS.md 是否存在
 * 2. UDS 標記區塊是否存在
 * 3. 標記區塊中列出的標準是否與 .standards/ 目錄一致
 *
 * @param cwd - 專案根目錄
 * @returns QualityGateStep 或 null（若無 AGENTS.md）
 */
export async function checkAgentsMdSync(
  cwd: string,
): Promise<{ step: QualityGateStep; driftedFiles?: string[] } | null> {
  const agentsMdPath = join(cwd, "AGENTS.md");
  const standardsDir = join(cwd, ".standards");

  let agentsMdContent: string;
  try {
    agentsMdContent = await readFile(agentsMdPath, "utf-8");
  } catch {
    return null; // No AGENTS.md — skip check
  }

  // Extract UDS marker block
  const startMarker = "<!-- UDS:STANDARDS:START -->";
  const endMarker = "<!-- UDS:STANDARDS:END -->";
  const startIdx = agentsMdContent.indexOf(startMarker);
  const endIdx = agentsMdContent.indexOf(endMarker);

  if (startIdx === -1 || endIdx === -1) {
    return {
      step: {
        name: "agents_md_check",
        command: "agents_md_check",
        passed: true,
        output: "AGENTS.md 存在但無 UDS 標記區塊，跳過檢查。",
      },
    };
  }

  // Parse listed standards from marker block
  const block = agentsMdContent.slice(startIdx + startMarker.length, endIdx);
  const listedFiles = new Set<string>();
  for (const match of block.matchAll(/`([^`]+\.ai\.yaml)`/g)) {
    listedFiles.add(match[1]);
  }

  // Read actual .standards/ directory
  let actualFiles: Set<string>;
  try {
    const entries = await readdir(standardsDir);
    actualFiles = new Set(entries.filter(f => f.endsWith(".ai.yaml")));
  } catch {
    return {
      step: {
        name: "agents_md_check",
        command: "agents_md_check",
        passed: false,
        output: ".standards/ 目錄不存在，但 AGENTS.md 有 UDS 標記區塊。請執行 uds init。",
      },
    };
  }

  // Compare
  const missing = [...actualFiles].filter(f => !listedFiles.has(f));
  const extra = [...listedFiles].filter(f => !actualFiles.has(f));
  const drifted = [...missing, ...extra];

  if (drifted.length === 0) {
    return {
      step: {
        name: "agents_md_check",
        command: "agents_md_check",
        passed: true,
        output: `AGENTS.md 與 .standards/ 同步（${actualFiles.size} 項標準）。`,
      },
    };
  }

  const details: string[] = [];
  if (missing.length > 0) details.push(`新增但未列入 AGENTS.md：${missing.join(", ")}`);
  if (extra.length > 0) details.push(`AGENTS.md 列出但已移除：${extra.join(", ")}`);

  return {
    step: {
      name: "agents_md_check",
      command: "agents_md_check",
      passed: false,
      output: `AGENTS.md drift detected:\n${details.join("\n")}\n建議執行 uds update 更新。`,
    },
    driftedFiles: drifted,
  };
}

/**
 * 前端設計合規性驗證常數
 *
 * 來源：UDS frontend-design-standards.ai.yaml（XSPEC-026 Phase 1）
 */

/** DESIGN.md 必填段落（支援 snake_case 和 kebab-case 兩種格式） */
const REQUIRED_DESIGN_SECTIONS = [
  ["visual_theme", "visual-theme"],
  ["color_palette", "color-palette"],
  ["typography"],
  ["component_styling", "component-styling"],
  ["layout_spacing", "layout-spacing"],
  ["design_guidelines", "design-guidelines"],
] as const;

/** 語義色彩必要 token（支援 snake_case 和 kebab-case 兩種格式） */
const REQUIRED_COLOR_TOKENS = [
  ["background"],
  ["surface"],
  ["primary_text", "primary-text"],
  ["muted_text", "muted-text"],
  ["accent"],
] as const;

/** design_guidelines.anti_patterns 建議最低條目數 */
const MIN_ANTI_PATTERN_COUNT = 5;

/**
 * 驗證前端設計合規性
 *
 * 非阻塞：DESIGN.md 不存在時回傳 null（純後端專案不應被 block）。
 * DESIGN.md 存在時驗證：
 * 1. 6 個必填段落是否完整
 * 2. 語義色彩 5 個必要 token 是否存在（warn）
 * 3. anti_patterns 條目數是否 ≥ 5（warn）
 *
 * @param cwd - 專案根目錄
 * @returns FrontendDesignCheckResult 或 null（若無 DESIGN.md）
 */
export async function checkFrontendDesignCompliance(
  cwd: string,
): Promise<FrontendDesignCheckResult | null> {
  const designMdPath = join(cwd, "DESIGN.md");

  let content: string;
  try {
    content = await readFile(designMdPath, "utf-8");
  } catch {
    // DESIGN.md 不存在 → 跳過（非前端專案不應被 block）
    return null;
  }

  const issues: string[] = [];
  const missingSections: string[] = [];
  const missingColorTokens: string[] = [];
  let antiPatternCount: number | undefined;

  // 1. 必填段落完整性檢查
  for (const aliases of REQUIRED_DESIGN_SECTIONS) {
    const found = aliases.some(alias =>
      // 符合 Markdown heading（## section_name）或 YAML key（section_name:）
      new RegExp(`(?:^#{1,6}\\s+${alias}\\b|^${alias}\\s*:)`, "im").test(content)
    );
    if (!found) {
      const primaryName = aliases[0];
      missingSections.push(primaryName);
    }
  }

  if (missingSections.length > 0) {
    issues.push(
      `缺少必填段落（${missingSections.length} 個）：${missingSections.join(", ")}\n` +
      `  → 請在 DESIGN.md 中補充上述段落（支援 ## section_name 或 section_name: 格式）`
    );
  }

  // 2. 語義色彩 token 檢查（warn）
  for (const aliases of REQUIRED_COLOR_TOKENS) {
    const found = aliases.some(alias =>
      new RegExp(`\\b${alias}\\s*[=:\\-]`, "i").test(content)
    );
    if (!found) {
      missingColorTokens.push(aliases[0]);
    }
  }

  if (missingColorTokens.length > 0) {
    issues.push(
      `語義色彩 token 不完整，缺少：${missingColorTokens.join(", ")}\n` +
      `  → 建議在 color_palette 段落中補充上述 token`
    );
  }

  // 3. anti_patterns 條目數檢查（warn）
  // 計算 anti_patterns 區塊中的清單條目（- 或 * 開頭的行）
  const antiPatternSection = content.match(
    /(?:anti[_-]patterns?)\s*[:\n]([^]*?)(?=\n#{1,6}\s|\n\w+[_-]\w+\s*:|$)/i
  );
  if (antiPatternSection) {
    const listItems = (antiPatternSection[1].match(/^\s*[-*+]\s+\S/gm) ?? []).length;
    antiPatternCount = listItems;
    if (listItems < MIN_ANTI_PATTERN_COUNT) {
      issues.push(
        `design_guidelines.anti_patterns 條目不足（目前 ${listItems} 條，建議 ≥ ${MIN_ANTI_PATTERN_COUNT} 條）\n` +
        `  → 請補充常見的前端設計反模式`
      );
    }
  }

  const passed = missingSections.length === 0;
  const outputLines: string[] = [];

  if (passed && issues.length === 0) {
    outputLines.push("DESIGN.md 前端設計合規性驗證通過。");
  } else {
    outputLines.push("DESIGN.md 前端設計合規性問題：");
    outputLines.push(...issues.map((issue, i) => `\n[${i + 1}] ${issue}`));
    if (passed) {
      outputLines.push("\n（必填段落完整，上述為 warning 項目）");
    } else {
      outputLines.push("\n請修正上述 error 項目（必填段落缺失），warning 項目建議補充。");
    }
  }

  return {
    step: {
      name: "frontend_design_check",
      command: "frontend_design_check",
      passed,
      output: outputLines.join(""),
    },
    missingSections: missingSections.length > 0 ? missingSections : undefined,
    missingColorTokens: missingColorTokens.length > 0 ? missingColorTokens : undefined,
    antiPatternCount,
  };
}

function buildFailResult(
  steps: QualityGateStep[],
  failedStep: QualityGateStep,
  evidence: VerificationEvidence[],
  task?: Task,
): QualityGateResult {
  const feedback = [
    `品質門檻「${failedStep.name}」失敗。`,
    `執行指令：${failedStep.command}`,
    `輸出：`,
    failedStep.output,
    "",
    "請修正上述問題後重試。",
  ].join("\n");

  return {
    passed: false,
    steps,
    feedback,
    evidence,
    ...(task?.spec_score != null && {
      score: task.spec_score,
      max_score: task.spec_max_score ?? (task.spec_score <= 10 ? 10 : 25),
    }),
  };
}
